### ISSUE TITLE
Improve Match Exactly notification experience

### ISSUE THREAD

**Issue by seadowg** [getodk/collect#4787]
Match Exactly notifications currently have two problems we've identified:

1. Clicking a failure notification will open the "Fill Blank Form" screen for the current project, which might not be the project that experienced the error
2. Failures happen pretty frequently - just under 20% syncs will result in an failure notification ([here](https://console.firebase.google.com/u/0/project/api-project-322300403941/analytics/app/android:org.odk.collect.android/events/~2Freport~2FMatchExactlySyncCompleted%3Ft%3D1628690834457&fpn%3D322300403941&swu%3D1&sgu%3D1&sus%3Dupgraded&params%3D_u..pageSize%253D25%2526_r..eventId%253DMatchExactlySyncCompleted&cs%3Dapp.m.events.detail&r%3Ddefault&g%3D1))

To alleviate these, we should make the following changes:

- [x] The notification should specify the project that failed to sync (clicking it should still just take the user to "Fill Blank Form" for now) in the subtext (fixed in https://github.com/getodk/collect/pull/4929)
- [x] Instead of showing a notification immediately, a failure (from any request, including form downloads) should result in a limited number of retries which is limited to a certain number of retries (most likely using `WorkManager`'s [Retry and Backoff](https://developer.android.com/topic/libraries/architecture/workmanager/how-to/define-work#retry_and_backoff_policy) and `WorkerParameters#getRunAttemptCount`) followed by a failure notification if the sync still hasn't succeeded
- [x] Update analytics calls to only measure failures that the user sees

We're not doing a lot to fix problem **1** here, but my thinking is we try and fix problem **2** as much as possible to make **1** less of a common experience. If we're right in that thinking, we should see the proportion of **visible** failures drop.

This can definitely be solved as multiple PRs.

**Comment by grzesiek2010** [comment]
In case of `Match Exactly` triggering new downloads maybe would make sense to solve some random failures but I'm afraid that that percent (of random failures) might be relatively small.

**Comment by lognaturel** [comment]
The random failures are things both @seadowg and I have experienced on test devices. I believe he has looked at server logs and confirmed requests are not making it to the server. I’ve had it happen on wifi in my office but I seem to recall he’s only seen it when walking around.

I think it definitely makes sense to retry in the case of requests not making it out to the server. It might make sense with some server response codes but not all. As always, I’m concerned about overloading servers.

**Comment by seadowg** [comment]
> I think it definitely makes sense to retry in the case of requests not making it out to the server. It might make sense with some server response codes but not all. As always, I’m concerned about overloading servers.

That's a great point. We probably only want to do this for `Unreachable` exceptions initially and see if that reduces the error rate significantly.

**Comment by grzesiek2010** [comment]
> We probably only want to do this for Unreachable exceptions initially and see if that reduces the error rate significantly.

That makes sense.

**Comment by grzesiek2010** [comment]
I think this issue is not consistent with https://github.com/getodk/collect/issues/4844. I'm talking about the second point:

>  Add "Show Details" button to notification that shows our new error view

in https://github.com/getodk/collect/pull/4925 which fixes https://github.com/getodk/collect/issues/4844 I implemented showing errors after clicking the notification (according to the issue) without any additional buttons. Shouldn't the behavior here be the same?

**Comment by seadowg** [comment]
> Shouldn't the behavior here be the same?

Good question. I think this came up in a discussion, but I should have written it down. 

For Match Exactly, I think clicking the notification should get them to the Fill Blank Form page where they can easily retry the sync (with the ♻️ button). That's the current flow and I felt like it's still better to prioritize them being able to retry over seeing error details. 

With auto update, I just retained the flow of getting straight to the error/success messages from the notification (but they will be more detailed now). We could make these consistent, but I don't think it's a big deal really. Making them consistent would also duplicate the problem that Match Exactly has where the user ends up on Fill Blank Form for a different project (which we will need to solve down the road probably). If you disagree, I think we can add an issue to add a `Show details` button to the auto update notifications and have the primary click take the user to Get Blank Form.

**Comment by grzesiek2010** [comment]
Today I started looking into this issue and playing with backoff policy. To be 100% sure the idea is to retry jobs that are triggered by `match exactly` (currently we don't retry at all if there is an exception we just display it) and throw exceptions but only if that exception is Unreachable exception.  In such a case we should try let's say 3 times and display a notification only after the last attempt. Do I understand everything correctly?

**Comment by seadowg** [comment]
>  In such a case we should try let's say 3 times and display a notification only after the last attempt. Do I understand everything correctly?

This sounds about right to me. Looking at the options we have in [WorkManager](https://developer.android.com/topic/libraries/architecture/workmanager/how-to/define-work#retries_backoff), I'm thinking we could use the `EXPONENTIAL` back off policy and set a minimum time of 60 seconds/1 minute? Given our fastest update interval is 15 mins, I think it makes sense to wait longer than the standard 10 seconds and then wait longer each time. 3 attempts feels right to me, but I have no logic to back it up really!

I'm also thinking we might want to retry on `SecurityError` as I've seen myself that seem to just be intermittent (device) connection problems.

`WorkerAdapter` already supports the retry stuff (based on returning `true` or `false` the work returned by `TaskSpec#getTask` so we could just add methods to `TaskSpec` that return the minimum retry time, max attempts etc

**Comment by grzesiek2010** [comment]
All what you said seems fine and is basically what I tested yesterday there is just one problem...

The whole logic responsible for downloading forms for `matchexactly` including handling exceptions (`FormSourceException` in this case) is hidden behind `FormsUpdater#matchFormsWithServer()`.  That method returns true if task is completed and false otherwise. As you said we could use that value to pass it to `WorkerAdapter` ([currently we always return true no mater what the result of `FormsUpdater#matchFormsWithServer()` is not to support any retries](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/backgroundwork/SyncFormsTaskSpec.java#L32)) and that would work fine (of course we would need to add backoff policy too) but...
You said we want to retry only as a result of some particular exceptions like `Unreachable`. If this is what we want to implement it's going to add some complexity (that I don't like) to our code. We would need maybe instead of returning simple true/false in `FormsUpdater#matchFormsWithServer()` return a pair of values something like `isCompleted`/`shouldRetry` (because `isCompleted` should be set to false as a response to any exception and `shouldRetry` only if it was `Unreachable`).

First you said that `Unreachable` should be the exception that should cause retrying and you also mentioned `SecurityError`, I thought about `ServerError` too.

So the thing to consider is: Would that be bad if we just retry no matter what the exception is so that we don't have to add that complexity to `FormsUpdater#matchFormsWithServer()`?

**Comment by seadowg** [comment]
> So the thing to consider is: Would that be bad if we just retry no matter what the exception is so that we don't have to add that complexity to FormsUpdater#matchFormsWithServer()?

Personally I do think we should go with only retrying for `Unreachable`/`SecurityError` as these are the only ones that can be caused by the device's access to the network rather than the server genuinely being in a bad state - not something we want to obscure. I think retrying on any error would be a good first step/PR though as you're right that it's certainly simpler to implement. I can pull out another task unless @lognaturel disagrees with me that we shouldn't release a version that retries on any error.

> We would need maybe instead of returning simple true/false in FormsUpdater#matchFormsWithServer() return a pair of values something like isCompleted/shouldRetry (because isCompleted should be set to false as a response to any exception and shouldRetry only if it was Unreachable).

I think returning a tristate type (using a sealed class or an enum) would solve this:

```kotlin
enum class Result {
    SUCCESS,
    INCOMPLETE,
    FAILURE
}
```

Here a `SUCCESS` would be any branch that currently returns `true`, an `INCOMPLETE` would be an `Unreachable` or `SecurityError` and a `FAILURE` would be any other error. It's then easy enough to pattern match on that in the `SyncFormsTaskSpec` and decide whether to return to `true` or `false` there.

**Comment by grzesiek2010** [comment]
Yes such a `Result` would also work but still it's adding complexity that only one client of `FormsUpdater#matchFormsWithServer()` needs - the second client is in `BlankFormsListViewModel` and there it doesn't matter what the exception was. Let's wait for @lognaturel.

**Comment by lognaturel** [comment]
For me the biggest pain point right now is when there's some kind of form error that gets entirely obscured. That really breaks the draft iteration process with Central in (relatively rare but still) cases where Collect errors out when parsing the form definition. Luckily that's mostly hitting office staff (as opposed to field staff) but I suspect it's wasting a lot of time. Just addressing that would be a huge first step and maybe that's its own PR.

The second biggest pain point is that the "everything is ok alarm" is sometimes showing that things aren't ok which can lead to unnecessary panic for field users. That's where the retry comes in. I don't mind retrying for all failure cases initially and even releasing that. That's because I believe the fixable failure cases are more likely to happen when iterating on the form which usually will require manual ♻️ update. 

Longer term, I think we might want to rethink putting up a notification in the case of a successful attempt to update. I was on a call with someone iterating on the form the other day and both of our phones kept making their notification sounds because we each had a few projects + the one we were iterating on. We did it this way to give users confirmation that the update succeeded, there just wasn't anything to pull. I don't know how well this is meeting that goal. Perhaps we could only show a notification when there's a failure or when the update attempt led to something actually being updated. That can be another issue.

**Comment by seadowg** [comment]
> For me the biggest pain point right now is when there's some kind of form error that gets entirely obscured. 

Are you meaning because we aren't bubbling up form parse errors from the individual form downloads? If so, yeah that is a different case than the one we're solving here (this issue description is focused on avoid intermittent errors). We should write something else up to cover that though! I'll add a note to talk about that in our next sync up.

>  I don't mind retrying for all failure cases initially and even releasing that. That's because I believe the fixable failure cases are more likely to happen when iterating on the form which usually will require manual ♻️ update.

Good point. I think long term it would still be better to have some errors show notifications immediately but if you're both preferring blanket retries for this version I'm happy with it. I'll edit the tasks in the issue description.

> Longer term, I think we might want to rethink putting up a notification in the case of a successful attempt to update.

Exactly Match Server doesn't/shouldn't show a notification on a successful sync. Was this using Previously Downloaded Only?

**Comment by lognaturel** [comment]
I'm a bit confused about what has happened here. I thought the error messages were the top priority and that was the first point described in https://github.com/getodk/collect/issues/4787#issuecomment-987116065. Fine to split it off, certainly!

> Exactly Match Server doesn't/shouldn't show a notification on a successful sync.

Definitely using match exactly. I'll pay better attention and file an issue if something doesn't seem right. I do tend to have projects with drafts that were abandoned from Central so maybe we both had those (they fail forever until deleted, at some point we should have a draft-specific experience). Whatever it was, there were lots of notifications from both sides.

**Comment by grzesiek2010** [comment]
Ok so we agreed that for now we should retry no matter what exception we have plus we should try like 3 times. Should we display a notification for every attempt if it fails or maybe wait for the last one?

**Comment by lognaturel** [comment]
> wait for the last one

Yes, wait for the last one. 👍

**Comment by seadowg** [comment]
> I'm a bit confused about what has happened here. I thought the error messages were the top priority and that was the first point

I'm confused as well. It looks like we broke that work out - there was a task to add a "Show details" button to the notification, but we removed this and then I presumably forgot to add the issue. My bad. That would have driven out increasing the granularity of our errors so that we're not swallowing the individual download ones. We can chat about that later we sync up - I'd like to talk about the iteration flow as well.
